File: examples/hello-world/gabriel-greeting-java/src/test/java/org/organicprogramming/gabriel/greeting/javaholon/api/CliTest.java
package org.organicprogramming.gabriel.greeting.javaholon.api;

import com.google.gson.Gson;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class CliTest {
    private static final Gson GSON = new Gson();

    @Test
    void runPrintsVersion() {
        ByteArrayOutputStream stdout = new ByteArrayOutputStream();
        ByteArrayOutputStream stderr = new ByteArrayOutputStream();

        int exitCode = Cli.run(new String[] { "version" }, new PrintStream(stdout), new PrintStream(stderr));

        assertEquals(0, exitCode);
        assertEquals(Cli.VERSION, stdout.toString(StandardCharsets.UTF_8).trim());
        assertEquals("", stderr.toString(StandardCharsets.UTF_8));
    }

    @Test
    void runPrintsHelp() {
        ByteArrayOutputStream stdout = new ByteArrayOutputStream();
        ByteArrayOutputStream stderr = new ByteArrayOutputStream();

        int exitCode = Cli.run(new String[] { "help" }, new PrintStream(stdout), new PrintStream(stderr));

        assertEquals(0, exitCode);
        assertTrue(stdout.toString(StandardCharsets.UTF_8).contains("usage: gabriel-greeting-java"));
        assertTrue(stdout.toString(StandardCharsets.UTF_8).contains("listLanguages"));
        assertEquals("", stderr.toString(StandardCharsets.UTF_8));
    }

    @Test
    void runRendersListLanguagesJson() {
        ByteArrayOutputStream stdout = new ByteArrayOutputStream();
        ByteArrayOutputStream stderr = new ByteArrayOutputStream();

        int exitCode = Cli.run(
                new String[] { "listLanguages", "--format", "json" },
                new PrintStream(stdout),
                new PrintStream(stderr));

        Map<?, ?> payload = GSON.fromJson(stdout.toString(StandardCharsets.UTF_8), Map.class);
        List<?> languages = (List<?>) payload.get("languages");

        assertEquals(0, exitCode);
        assertEquals(56, languages.size());
        Map<?, ?> first = (Map<?, ?>) languages.get(0);
        assertEquals("en", first.get("code"));
        assertEquals("English", first.get("name"));
        assertEquals("", stderr.toString(StandardCharsets.UTF_8));
    }

    @Test
    void runRendersSayHelloText() {
        ByteArrayOutputStream stdout = new ByteArrayOutputStream();
        ByteArrayOutputStream stderr = new ByteArrayOutputStream();

        int exitCode = Cli.run(
                new String[] { "sayHello", "Alice", "fr" },
                new PrintStream(stdout),
                new PrintStream(stderr));

        assertEquals(0, exitCode);
        assertEquals("Bonjour Alice", stdout.toString(StandardCharsets.UTF_8).trim());
        assertEquals("", stderr.toString(StandardCharsets.UTF_8));
    }

    @Test
    void runDefaultsSayHelloToEnglishJson() {
        ByteArrayOutputStream stdout = new ByteArrayOutputStream();
        ByteArrayOutputStream stderr = new ByteArrayOutputStream();

        int exitCode = Cli.run(
                new String[] { "sayHello", "--json" },
                new PrintStream(stdout),
                new PrintStream(stderr));

        Map<?, ?> payload = GSON.fromJson(stdout.toString(StandardCharsets.UTF_8), Map.class);

        assertEquals(0, exitCode);
        assertEquals("Hello Mary", payload.get("greeting"));
        assertEquals("English", payload.get("language"));
        assertEquals("en", payload.get("langCode"));
        assertEquals("", stderr.toString(StandardCharsets.UTF_8));
    }
}

File: examples/hello-world/gabriel-greeting-kotlin/src/test/kotlin/org/organicprogramming/gabriel/greeting/kotlinholon/api/CliTest.kt
package org.organicprogramming.gabriel.greeting.kotlinholon.api

import com.google.gson.Gson
import java.io.ByteArrayOutputStream
import java.io.PrintStream
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class CliTest {
    private val gson = Gson()

    @Test
    fun runPrintsVersion() {
        val stdout = ByteArrayOutputStream()
        val stderr = ByteArrayOutputStream()

        val exitCode = Cli.run(arrayOf("version"), PrintStream(stdout), PrintStream(stderr))

        assertEquals(0, exitCode)
        assertEquals(Cli.VERSION, stdout.toString().trim())
        assertEquals("", stderr.toString())
    }

    @Test
    fun runPrintsHelp() {
        val stdout = ByteArrayOutputStream()
        val stderr = ByteArrayOutputStream()

        val exitCode = Cli.run(arrayOf("help"), PrintStream(stdout), PrintStream(stderr))

        assertEquals(0, exitCode)
        assertTrue(stdout.toString().contains("usage: gabriel-greeting-kotlin"))
        assertTrue(stdout.toString().contains("listLanguages"))
        assertEquals("", stderr.toString())
    }

    @Test
    fun runRendersListLanguagesJson() {
        val stdout = ByteArrayOutputStream()
        val stderr = ByteArrayOutputStream()

        val exitCode = Cli.run(arrayOf("listLanguages", "--format", "json"), PrintStream(stdout), PrintStream(stderr))
        val payload = gson.fromJson(stdout.toString(), Map::class.java)
        val languages = payload["languages"] as List<*>
        val first = languages.first() as Map<*, *>

        assertEquals(0, exitCode)
        assertEquals(56, languages.size)
        assertEquals("en", first["code"])
        assertEquals("English", first["name"])
        assertEquals("", stderr.toString())
    }

    @Test
    fun runRendersSayHelloText() {
        val stdout = ByteArrayOutputStream()
        val stderr = ByteArrayOutputStream()

        val exitCode = Cli.run(arrayOf("sayHello", "Alice", "fr"), PrintStream(stdout), PrintStream(stderr))

        assertEquals(0, exitCode)
        assertEquals("Bonjour Alice", stdout.toString().trim())
        assertEquals("", stderr.toString())
    }

    @Test
    fun runDefaultsSayHelloToEnglishJson() {
        val stdout = ByteArrayOutputStream()
        val stderr = ByteArrayOutputStream()

        val exitCode = Cli.run(arrayOf("sayHello", "--json"), PrintStream(stdout), PrintStream(stderr))
        val payload = gson.fromJson(stdout.toString(), Map::class.java)

        assertEquals(0, exitCode)
        assertEquals("Hello Mary", payload["greeting"] as String)
        assertEquals("English", payload["language"] as String)
        assertEquals("en", payload["langCode"] as String)
        assertEquals("", stderr.toString())
    }
}
